### PR TITLE
JIT: Deduplicate CodeGen::genJumpTable

### PR DESCRIPTION
**Pull request by yurai007**
Moving same code used across all architectures to common codegen should simplify things and reduce possibility of future regressions as noted in #98992.

Part of https://github.com/dotnet/runtime/issues/84834
cc @wscho77 @HJLeee @clamp03 @JongHeonChoi @t-mustafin @gbalykov @viewizard @ashaurtaev @brucehoult @sirntar @yurai007 @Bajtazar @bartlomiejko @rzsc @tomeksowi @amanasifkhalid

### CODE REVIEW

**Review by Bajtazar — APPROVED** [review]

**Review by gbalykov — APPROVED** [review]

**Review by sirntar — APPROVED** [review]

File: src/coreclr/jit/codegencommon.cpp [hunk 1]
```diff
@@ -6452,6 +6452,46 @@ void CodeGen::genFnProlog()
...
+// Arguments:
+//    treeNode - the GT_JMPTABLE node
+//
+// Return Value:
+//    base offset to jump table
+//
+// Assumption:
+//    The current basic block in process ends with a switch statement
+//
+unsigned CodeGen::getBaseOffsetOfGeneratedJumpTable(GenTree* treeNode)
```
Comment (jakobbotsch) [inline src/coreclr/jit/codegencommon.cpp:6467]: Nit: the name makes it seem like the jump table has already been generated, but this function is what emits it.
I would call it `genEmitJumpTable` instead.

**Review by tomeksowi — APPROVED** [review]

**Review by amanasifkhalid — APPROVED** [review]
LGTM. Could you please run jitformat locally to fix the formatting issues? Thanks!

File: src/coreclr/jit/codegenarm.cpp [hunk 2]
```diff
@@ -647,30 +647,7 @@ void CodeGen::genTableBasedSwitch(GenTree* treeNode)
...
 void CodeGen::genJumpTable(GenTree* treeNode)
 {
-    noway_assert(compiler->compCurBB->KindIs(BBJ_SWITCH));
-    assert(treeNode->OperGet() == GT_JMPTABLE);
-
-    unsigned   jumpCount = compiler->compCurBB->GetSwitchTargets()->bbsCount;
-    FlowEdge** jumpTable = compiler->compCurBB->GetSwitchTargets()->bbsDstTab;
-    unsigned   jmpTabBase;
-
-    jmpTabBase = GetEmitter()->emitBBTableDataGenBeg(jumpCount, false);
```
Comment (bartlomiejko) [inline src/coreclr/jit/codegenarm.cpp:657]: Not sure if this is important, but here you have `false `in the argument of calling

> jmpTabBase = GetEmitter()->emitBBTableDataGenBeg(jumpCount, false);

in the rest calls and new method you have `true`

> jmpTabBase = GetEmitter()->emitBBTableDataGenBeg(jumpCount, true);

why is this difference, is it needed and can it be changed?

**Review by bartlomiejko — COMMENTED** [review]
Please check comment.

File: src/coreclr/jit/codegenarm.cpp (hunk 2 above)
Comment (yurai007) [reply src/coreclr/jit/codegenarm.cpp:657]: Yes, I think it is :) Thanks for catch.

Comment (tomeksowi) [reply src/coreclr/jit/codegenarm.cpp:657]: BTW, does this argument matter on a 32-bit platform?
https://github.com/dotnet/runtime/blob/0a5418d7e7ae7f7653a32004c808881af5275469/src/coreclr/jit/emit.cpp#L7885-L7935

File: src/coreclr/jit/codegen.h [hunk 3]
```diff
@@ -1260,6 +1260,7 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
     void genCodeForInitBlkLoop(GenTreeBlk* initBlkNode);
     void genCodeForInitBlkRepStos(GenTreeBlk* initBlkNode);
     void genCodeForInitBlkUnroll(GenTreeBlk* initBlkNode);
+    unsigned genEmitJumpTable(GenTree* treeNode, bool relativeAddr);
     void genJumpTable(GenTree* tree);
     void genTableBasedSwitch(GenTree* tree);
 #if defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
```
Comment (jakobbotsch) [inline src/coreclr/jit/codegen.h:1266]: It seems like the file mode is being changed on all these files (I see 100644 -> 100755) on them. Can you revert that change?

**Review by jakobbotsch — APPROVED** [review]
Thank you!

**Pull request merged**